docs: Add doc comments and README descriptions about WebGPU MToon stuff

0b5vr opened the pull request:
Changes of imports in this PR are all type imports and only used for doc comments

I have to add support for v0CompatShade to the WebGPU one later

I think it's high time to remove the v1 notice on the root README

I thought that we could specify more simpler option like `nodeMaterial: true` to the option of `VRMLoaderPlugin` while writing these docs

### Points need review

- [ ] Do you think it's easy to follow when you try to use three-vrm in WebGPURenderer?

0b5vr committed:
docs: Add doc comments and README descriptions about WebGPU MToon stuff

changes of imports in this commit are all type imports and only used for doc comments

I have to add support for v0CompatShade to the WebGPU one later

I think it's high time to remove the v1 notice on the root README

I thought that we could specify more simpler option like `nodeMaterial: true` to the option of `VRMLoaderPlugin` while writing these docs

yue4u commented on README.md:
```diff
@@ -138,6 +128,32 @@ loader.load(
+  return new VRMLoaderPlugin(parser, {
+
+    // Specify the MToon loader plugin to use in the VRMLoaderPlugin instance
+    mtoonMaterialPlugin,
```
Question: Can this be detected inside three-vrm? Should we mention `renderer.isWebGPURenderer`?

0b5vr replied:
I think `VRMLoaderPlugin` doesn't have an access to `renderer`.

yue4u replied:
> Can this be detected inside three-vrm
So this is no.

> Should we mention

I mean if the renderer may be created before `VRMLoaderPlugin` registration and exist somewhere in current execution context, user can figure out what to use via `isWebGPURenderer`.

0b5vr replied:
Renderer and VRMLoaderPlugin should be loosely coupled I think. What does the code you expected look like?

yue4u replied:
Something like
```ts
const mtoonMaterialPlugin = renderer.isWebGPURenderer === true 
  ? new MToonNodeMaterialLoaderPlugin(parser)
  : undefined
```

or just a comment

```
// assert renderer.isWebGPURenderer is true
```

I know this is totally optional but may provide some useful info to users.

0b5vr replied:
I see I see.

https://github.com/pixiv/three-vrm/pull/1420/files#r1642356443 👈 Does this look good?

yue4u commented on packages/three-vrm-materials-mtoon/src/MToonMaterialLoaderPlugin.ts:
```diff
@@ -1,19 +1,26 @@
 import type { MToonMaterialDebugMode } from './MToonMaterialDebugMode';
 import { GLTF as GLTFSchema } from '@gltf-transform/core';
 import { MToonMaterial } from './MToonMaterial';
+import { MToonNodeMaterialLoaderPlugin } from './nodes/MToonNodeMaterialLoaderPlugin';
```
```suggestion
import type { MToonNodeMaterialLoaderPlugin } from './nodes/MToonNodeMaterialLoaderPlugin';
```

yue4u commented on packages/three-vrm-materials-mtoon/src/nodes/MToonNodeMaterial.ts:
```diff
@@ -43,6 +44,14 @@ import { mtoonParametricRim } from './mtoonParametricRim';
 // import phongLightingModel from 'three/addons/nodes/functions/PhongLightingModel.js';
 // import { float } from '../shadernode/ShaderNode.js';
 
+/**
+ * MToon is a material specification that has various features.
+ * The spec and implementation are originally founded for Unity engine and this is a port of the material.
+ *
+ * This material is a NodeMaterial variant of {@link MToonMaterial}.
+ *
+ * See: https://github.com/Santarh/MToon
+ */
```
👍

0b5vr committed:
fix: MToonMaterialLoaderPlugin only needs MToonNodeMaterialLoaderPlugin as a type import

Co-authored-by: yue <yue@pixiv.co.jp>

0b5vr committed:
docs: proofread the WebGPU section of the README

Google Translate turned my Japanese into English so nicely

0b5vr commented on README.md:
```diff
@@ -138,6 +128,32 @@ loader.load(
+loader.register((parser) => {
+
+  // create a WebGPU compatible MToon loader plugin
+  const mtoonMaterialPlugin = new MToonNodeMaterialLoaderPlugin(parser);
```
```suggestion
  const mtoonMaterialPlugin = renderer.isWebGPURenderer === true
    ? new MToonNodeMaterialLoaderPlugin(parser)
    : undefined; // fallbacks to the default `MToonMaterialLoaderPlugin`
```

0b5vr replied:
This is kinda redundant considering that the document already says that it's for `WebGPURenderer`.

0b5vr committed:
docs: MToon, add notice about supported Three.js version when using MToonNodeMaterial

I also added a warning that we may break compatibility more frequently when using NodeMaterial

ke456-png commented on packages/three-vrm-materials-mtoon/README.md:
```diff
@@ -10,9 +10,11 @@ MToon (toon material) module for @pixiv/three-vrm
+To use three-vrm with WebGPURenderer, specify the WebGPU-compatible `MToonNodeMaterialLoaderPlugin` for the `mtoonMaterialPlugin` option of `VRMLoaderPlugin`.
 
-To use MToon with `VRMLoaderPlugin` in WebGPURenderer, specify the WebGPU-compatible `MToonNodeMaterialLoaderPlugin` for the `mtoonMaterialPlugin` option of `VRMLoaderPlugin`.
+`MToonNodeMaterial` only supports Three.js r161 or later.
+The NodeMaterial system of Three.js is still under development, so we may break compatibility with older versions of Three.js more frequently than other parts of three-vrm.
```
Thanks!
